compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt

```diff
@@ -168,3 +231,94 @@ public fun AnvilCompilationExceptionTypReference(
     )
   }
 }
+
+private fun resolveGenericTypeReference(
+  implementingClass: ClassReference.Psi,
+  declaringClass: ClassReference,
+  parameterName: String
+): Psi? {
+  // If the class/interface declaring the generic is the receiver class,
+  // then the generic hasn't been set to a concrete type and can't be resolved.
+  if (implementingClass == declaringClass) return null
+
+  // Used to determine which parameter to look at in a KtTypeArgumentList.
+  val indexOfType = declaringClass.indexOfTypeParameter(parameterName)
+
+  // Find where the supertype is actually declared by matching the FqName of the
+  // SuperTypeListEntry to the type which declares the generic we're trying to resolve.
+  // After finding the SuperTypeListEntry, just take the TypeReference from its type
+  // argument list.
+  val resolvedTypeReference = implementingClass.superTypeListEntryOrNull(declaringClass.fqName)
+    ?.typeReference
+    ?.typeElement
+    ?.getChildOfType<KtTypeArgumentList>()
+    ?.arguments
+    ?.get(indexOfType)
+    ?.typeReference
+
+  return if (resolvedTypeReference != null) {
+    // This will check that the type can be imported.
+    val resolvedDeclaringClass = resolvedTypeReference.containingClass()
+      ?.toClassReference(implementingClass.module)
+      ?: return null
+
+    resolvedTypeReference
+      .toTypeReference(resolvedDeclaringClass)
+      .resolveTypeReference(implementingClass)
+  } else {
+    null
+  }
```

Comment:
I think you can simplify this as 

```suggestion
  return resolvedTypeReference
      ?.containingClass()
      ?.toClassReference(implementingClass.module)
      ?.let { resolvedDeclaringClass ->
          resolvedTypeReference
            .toTypeReference(resolvedDeclaringClass)
            .resolveTypeReference(implementingClass)
      }
```

Reply:
Yes!